include/bluetooth/audio/hap.h

```diff
@@ -0,0 +1,275 @@
+	void (*disconnected)(struct bt_hap *hap, int reason);
+
+	/**
+	 * @brief A profile instance has been deleted.
+	 *
+	 * This callback notifies the application that a profile
+	 * instance has been removed. This may be called when the bond has
+	 * been removed or @ref bt_hap_delete has been called.
+	 *
+	 * @param hap Profile instance object.
+	 */
+	void (*deleted)(struct bt_hap *hap);
+
+	/**
+	 * @brief An active preset has been changed.
+	 *
+	 * This callback notifies the application when the active preset or
+	 * it's name has been changed.
+	 * If @kconfig{BT_HAP_HARC_PRESET_CACHE_SIZE} is 0 the preset name will
+	 * not be set, but if @kconfig{BT_HAP_HARC_AUTO_READ_ACTIVE_PRESET}
+	 * option is enabled, the callback will be called again on preset read
+	 * complete.
+	 *
+	 * @param hap Profile instance object.
+	 * @param preset Preset record object.
+	 */
+	void (*active_preset_changed)(struct bt_hap *hap,
+				      const struct bt_has_preset *preset);
+
+	/**
+	 * @brief A cached preset list has been updated.
+	 *
+	 * This callback notifies the application when the list of preset
+	 * records changed i.e. record has been added/removed, changed name or
+	 * become available.
+	 * The user can call @ref bt_hap_preset_foreach to obtain the current
+	 * presets.
+	 *
+	 * @param hap Profile instance object.
+	 */
+	void (*preset_cache_updated)(struct bt_hap *hap);
+};
+
+/**
+ * @def BT_HAP_CB_DEFINE
+ *
+ * @brief Register a callback structure for Hearing Access Profile events.
+ *
+ * @param _name Name of callback structure.
+ */
+#define BT_HAP_CB_DEFINE(_name)							\
+	static const STRUCT_SECTION_ITERABLE(bt_hap_cb,				\
+					     _CONCAT(bt_hap_cb_, _name))
+
+/**
+ * @brief Create profile instance to remote Hearing Access Service
+ *
+ * This will start a GATT discovery to setup handles and subscriptions.
+ * This shall be called once before any other actions can be executed for the
+ * peer device, and the @ref bt_hap_cb connected callback will notify
+ * when it is possible to start remote operations.
+ * The profile instance is removed when the bond is removed or when @ref
+ * bt_hap_delete is called.
+ * 
+ * @param conn The connection to initialize the profile for.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_new(struct bt_conn *conn);
+
+/**
+ * @brief Delete profile instance
+ *
+ * This will remove stored profile instance of remote Hearing Aid Service.
+ *
+ * @param hap Profile instance object.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_delete(struct bt_hap *hap);
+
+/**
+ * @brief Iterate through all existing profile instances.
+ *
+ * Iterate profile instances of remote Hearing Access Services.
+ *
+ * @param func Function to call for each profile instance.
+ * @param user_data Data to pass to the callback function.
+ */
+void bt_hap_foreach(bool (*func)(const struct bt_hap *hap, void *user_data),
+		    void *user_data);
```

Comment:
While this may be useful, I don't see any use case for it. 

Can you give me an example of when/why this is useful?

Reply:
User can call bt_hap_foreach to e.g. later call bt_hap_delete to clean up HAP persistent storage

Reply:
But given that the user already should have all the references to the `bt_hap` struct from calling `bt_hap_new`, this sounds unneeded. 
If we have multiple applications (unlikely for this case) that have called `bt_hap_new`, deleting instances via `bt_hap_foreach` will be dangerous, as you might delete instances of other applications.

Reply:
OK, will remove that

Reply:
OK. 

I think the foreach makes more sense if we have global callbacks like we do for `bt_conn` :)